File: src/main/java/com/timepath/hl2/io/demo/DemoHeader.java
package com.timepath.hl2.io.demo;

import com.timepath.DataUtils;
import com.timepath.io.struct.StructField;

import java.nio.ByteBuffer;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @author TimePath
 */
public class DemoHeader {

    private static final Logger LOG        = Logger.getLogger(DemoHeader.class.getName());
    private static final int    MAX_OSPATH = 260;
    @StructField(index = 0, limit = 8)
    public String head;
    @StructField(index = 1)
    public int    demoProtocol;
    @StructField(index = 2)
    public int    networkProtocol;
    @StructField(index = 3, limit = MAX_OSPATH)
    public String serverName;
    @StructField(index = 4, limit = MAX_OSPATH)
    public String clientName;
    @StructField(index = 5, limit = MAX_OSPATH)
    public String mapName;
    @StructField(index = 6, limit = MAX_OSPATH)
    public String gameDirectory;
    @StructField(index = 7)
    public float  playbackTime;
    @StructField(index = 8)
    public int    ticks;
    @StructField(index = 9)
    public int    frames;
    @StructField(index = 10)
    public int    signonLength;

    static DemoHeader parse(ByteBuffer slice) {
        DemoHeader h = new DemoHeader();
        h.head = DataUtils.getText(DataUtils.getSlice(slice, 8));
        if(!HL2DEM.HEADER.equals(h.head)) {
            LOG.log(Level.WARNING, "Unexpected header");
        }
        h.demoProtocol = slice.getInt();
        if(h.demoProtocol != HL2DEM.DEMO_PROTOCOL) {
            LOG.log(Level.WARNING, "Unknown demo version {0}", h.demoProtocol);
        }
        h.networkProtocol = slice.getInt();
        LOG.log(Level.INFO, "Network protocol: {0}", h.networkProtocol);
        ByteBuffer serverNameBuffer = DataUtils.getSlice(slice, MAX_OSPATH);
        h.serverName = DataUtils.getText(serverNameBuffer, true);
        LOG.log(Level.INFO, "Server: {0}", h.serverName);
        ByteBuffer clientNameBuffer = DataUtils.getSlice(slice, MAX_OSPATH);
        h.clientName = DataUtils.getText(clientNameBuffer, true);
        LOG.log(Level.INFO, "Client: {0}", h.clientName);
        ByteBuffer mapNameBuffer = DataUtils.getSlice(slice, MAX_OSPATH);
        h.mapName = DataUtils.getText(mapNameBuffer, true);
        LOG.log(Level.INFO, "Map: {0}", h.mapName);
        ByteBuffer gameDirectoryBuffer = DataUtils.getSlice(slice, MAX_OSPATH);
        h.gameDirectory = DataUtils.getText(gameDirectoryBuffer, true);
        LOG.log(Level.INFO, "Game: {0}", h.gameDirectory);
        h.playbackTime = slice.getFloat();
        LOG.log(Level.INFO, "Playback time: {0}", h.playbackTime);
        h.ticks = slice.getInt();
        LOG.log(Level.INFO, "Ticks: {0}", h.ticks);
        h.frames = slice.getInt();
        LOG.log(Level.INFO, "Frames: {0}", h.frames);
        h.signonLength = slice.getInt();
        LOG.log(Level.INFO, "Signon length: {0}", h.signonLength);
        return h;
    }
}

File: src/main/kotlin/com/timepath/hl2/io/demo/DemoHeader.kt
package com.timepath.hl2.io.demo

import com.timepath.DataUtils
import com.timepath.Logger
import com.timepath.io.struct.StructField
import com.timepath.with
import java.nio.ByteBuffer

private val MAX_OSPATH = 260

public class DemoHeader(
        StructField(0, limit = 8) var head: String,
        StructField(1) var demoProtocol: Int,
        StructField(2) var networkProtocol: Int,
        StructField(3, limit = MAX_OSPATH) var serverName: String,
        StructField(4, limit = MAX_OSPATH) var clientName: String,
        StructField(5, limit = MAX_OSPATH) var mapName: String,
        StructField(6, limit = MAX_OSPATH) var gameDirectory: String,
        StructField(7) var playbackTime: Float,
        StructField(8) var ticks: Int,
        StructField(9) var frames: Int,
        StructField(10) var signonLength: Int
) {
    companion object {

        private val LOG = Logger()

        fun parse(slice: ByteBuffer): DemoHeader {
            val head = DataUtils.getText(DataUtils.getSlice(slice, 8))
            if (HL2DEM.HEADER != head) {
                LOG.warning { "Unexpected header" }
            }
            val demoProtocol = slice.getInt()
            if (demoProtocol != HL2DEM.DEMO_PROTOCOL) {
                LOG.warning { "Unknown demo version ${demoProtocol}" }
            }
            val networkProtocol = slice.getInt()
            val serverName = DataUtils.getText(DataUtils.getSlice(slice, MAX_OSPATH), true)
            val clientName = DataUtils.getText(DataUtils.getSlice(slice, MAX_OSPATH), true)
            val mapName = DataUtils.getText(DataUtils.getSlice(slice, MAX_OSPATH), true)
            val gameDirectory = DataUtils.getText(DataUtils.getSlice(slice, MAX_OSPATH), true)
            val playbackTime = slice.getFloat()
            val ticks = slice.getInt()
            val frames = slice.getInt()
            val signonLength = slice.getInt()
            LOG.info { "Network protocol: ${networkProtocol}" }
            LOG.info { "Server: ${serverName}" }
            LOG.info { "Client: ${clientName}" }
            LOG.info { "Map: ${mapName}" }
            LOG.info { "Game: ${gameDirectory}" }
            LOG.info { "Playback time: ${playbackTime}" }
            LOG.info { "Ticks: ${ticks}" }
            LOG.info { "Frames: ${frames}" }
            LOG.info { "Signon length: ${signonLength}" }
            return DemoHeader(
                    head = head,
                    demoProtocol = demoProtocol,
                    networkProtocol = networkProtocol,
                    serverName = serverName,
                    clientName = clientName,
                    mapName = mapName,
                    gameDirectory = gameDirectory,
                    playbackTime = playbackTime,
                    ticks = ticks,
                    frames = frames,
                    signonLength = signonLength
            )
        }
    }
}
